magma/lib/magma/loader.rb

```diff
@@ -134,6 +136,91 @@ def explicit_revision_exists?(revisions:, parent_model:, parent_record_name:, ch
+      #
+      # The identifiers DO NOT need to exist in the identifier table.
+
+      return {} if @grammar.nil?
+      grammar_decomposed = @grammar.decompose(identifier)
+      return {} if grammar_decomposed.nil?
+
+      # First we gather all available parent models and identifiers associated with the identifier.
+      # If someone has created a well formatted grammar, with proper identifiers,
+      # there MAY be a model hierarchy chain we can infer
+
+      project = Magma.instance.get_project(@project_name)
+      available_models = project.models.keys & grammar_decomposed[:rules].keys.map(&:to_sym)
+
+      chain = {}
+      available_models.each do |model_name|
+        model = project.models[model_name]
+        chain[model_name] = {
+            model: model,
+            identifier_name: model.identity.name,
+            identifier: grammar_decomposed[:rules][model_name.to_s][:name],
+            parent_model_name: model.parent_model_name
+          }
+      end
+
+      # Then we check to make sure a valid hierarchy chain exists
+      # We check that each parent_model is in the list of models available
+      arr = chain.to_a
+      valid = arr.all? do |el|
+        model_name = el[0]
+        model_name == :project || arr.any? do |other|
+          candidate_model_name = other[0]
+          el[1][:parent_model_name] == candidate_model_name
+        end
+      end
+
+      valid ? chain : {}
+
+    end
+
+    def autolink_parent_identifiers
+      # If a project is configured with the 'identifier' flag, anytime a record with an identifier is created or updated,
+      # we can implicitly infer what their parents must be via find_parent_models(). We then update the record and its revision
+      # to include its foreign key, and do this up the model hierarchy.
+      #
+      # This method handles a very specific update:
+      # - The incoming update does NOT explicitly reference a parent/child's foreign key.
+      # - It doesn't matter what type of attributes we are updating
+
+      # It is also important to note that this method is OPTIMISTIC. That is, even though we've inferred a model
+      # hierarchy through a grammar, the identifiers may not exist in the db, and we will only know once we call validate().
+      # Ideally we would validate() before we do this work, however the control flow of this classes makes it hard to do so.
+      # TODO: eventually refactor to address the above comment.
+      #
+      gnomon_mode = Magma::Flags::GNOMON_MODE
+      flag_value = @flags[gnomon_mode[:name]]
+
+      @records.each do |model, record_set|
+        record_set.each do |record_name, record|
+
+          next unless flag_value == gnomon_mode[:identifier] and not record.includes_parent_record?
+
+          # Attempt to find parent models
+          parent_models = find_parent_models(record_name)
+          next if parent_models.empty?
+
+          parent_models.each do |model_name, hash|
```

Comment:
might call this `model_info` or something instead of `hash`

Reply:
ya I like that